### PR TITLE
[Enhancement] enable rewrite sum by associative rule

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

you can see the motivation in RewriteSumByAssociativeRule.java
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                return newAggRef;
+            }
+        }
+
+        @Override
+        public ScalarOperator visitVariableReference(ColumnRefOperator op, Void context) {
+            if (oldAggregations.containsKey(op)) {
+                CallOperator aggFunction = oldAggregations.get(op);
+                if (aggFunction.isAggregate() && !aggFunction.isDistinct()
+                        && aggFunction.getFnName().equals("sum") &&
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:300]: use FunctionSet.SUM instead of literal

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+                    if (GlobalStateMgr.getCurrentState().getFunction(
+                            possibleAggregateFunction, Function.CompareMode.IS_IDENTICAL) != null) {
+                        return rewriteCastForAggExpr(fromOperator);
+                    }
+                }
+            }
+            return expr;
+        }
+
+        private ScalarOperator createNewAggFunction(ScalarOperator arg0, ScalarOperator arg1, Type returnType) {
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:197]: why we need this function ?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 3]
```diff
@@ -749,6 +750,9 @@ public class SessionVariable implements Serializable, Writable, Cloneable {
     @VarAttr(name = ENABLE_SORT_AGGREGATE)
     private boolean enableSortAggregate = false;
 
+    @VarAttr(name = DISABLE_REWRITE_SUM_BY_ASSOCIATIVE_RULE)
+    private boolean disableRewriteSumByAssociativeRule = false;
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:754]: use enable instead of disable

**Review by stdpain — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+import java.util.Map;
+
+// For aggregate functions like `sum(expr + const)`, we can rewrite it to `sum(expr) + const * count()`
+// to reduce the overhead of a project calculation.
+public class RewriteSumByAssociativeRule extends TransformationRule {
+    public RewriteSumByAssociativeRule() {
+        super(RuleType.TF_REWRITE_SUM_BY_ASSOCIATIVE_RULE,
+                Pattern.create(OperatorType.LOGICAL_PROJECT)
+                        .addChildren(Pattern.create(OperatorType.LOGICAL_AGGR)
+                                .addChildren(Pattern.create(OperatorType.LOGICAL_PROJECT))));
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:56]:
```suggestion
                                .addChildren(Pattern.create(OperatorType.PATTERN_LEAF))));
```

Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:56]:
```suggestion
        super(RuleType.TF_REWRITE_SUM_BY_ASSOCIATIVE_RULE,
                Pattern.create(OperatorType.LOGICAL_AGGR)
                                .addChildren(Pattern.create(OperatorType.PATTERN_LEAF)));
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    }
+
+    @Override
+    public boolean check(final OptExpression input, OptimizerContext context) {
+        return !context.getSessionVariable().isDisableRewriteSumByAssociativeRule();
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        // let's use sum(id_int +1) as an example, suppose id_int is INT
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:66]: support other arithmetic expr? like: sum(id_int - 1), sum(id_int * 2) ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        rewriteSumVisitor.newAggregations.putAll(rewriteSumVisitor.reservedAggregations);
+        newPreAggProjections.putAll(rewriteSumVisitor.oldPreAggProjections);
+
+        LogicalProjectOperator newPostAggProjectOperator =
+                new LogicalProjectOperator(newPostAggProjections);
+
+        LogicalAggregationOperator newAggOperator = new LogicalAggregationOperator(
+                aggregationOperator.getType(),
+                aggregationOperator.getGroupingKeys(), rewriteSumVisitor.newAggregations);
+        newAggOperator.setProjection(aggregationOperator.getProjection());
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:123]: use project node, not projection

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+        OptExpression newPreAggProjectOpt = OptExpression.create(new LogicalProjectOperator(newPreAggProjections),
+                input.getInputs().get(0).getInputs().get(0).getInputs());
+        newPreAggProjectOpt.setLogicalProperty(new LogicalProperty(new ColumnRefSet(newPreAggProjections.keySet())));
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:127]: don't need set logical property

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+         * ->  LogicalAggregation {type=GLOBAL ,aggregations={28: sum=sum(27: expr)} ,groupKeys=[]}
+         *   ->  LogicalProjectOperator {projection=[add(cast(1: id_int as bigint(20)), 1)]}
+         *
+         * After
+         *
+         * LogicalProjectOperator {projection=[add(30: sum, multiply(31: count, 1))]
+         * ->  LogicalAggregation {type=GLOBAL ,aggregations={30: sum=sum(29: id_int), 31: count=count()} ,groupKeys=[]}
+         *   ->  LogicalProjectOperator {projection=[add(cast(1: id_int as bigint(20)), 1), 1: id_int]}
+         */
+        LogicalProjectOperator postAggProjectOperator = (LogicalProjectOperator) input.getOp();
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:80]: if there is no project operator resides above agg operator, can this optimization be applied?
so in such a situation, a trivial project operation always is interpolated above agg operator？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java (hunk 1 above)
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:300]: can we support other agg functions, for examples:

avg(v+c) = avg(v)+c
avg (distinct v+c) = avg(distinct v) + c
max(v+c) = max(v) +c
min(v+c) = min(v) +c
count(v + c) = count(v)
count(distinct v+c) = count(distinct v)

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 3 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:754]: for high-cardinality group-by agg,  add another count agg functions, in the 1st AggregateStreamingOperator,  serialized overhead would increase significantly, should we testify these scenarios, if so ,we can prevent these rewrite.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java (hunk 5 above)
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:85]: > support other arithmetic expr? like: sum(id_int - 1), sum(id_int * 2) ?

currently only support sum(expr + const) and sum(expr -const), I think we can use this idea to rewrite sum(expr * const), I will do it later

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        // try to rewrite aggregate function top-down
+        for (Map.Entry<ColumnRefOperator, ScalarOperator> postAggProjection :
+                postAggProjectOperator.getColumnRefMap().entrySet()) {
+            ColumnRefOperator columnRefOperator = postAggProjection.getKey();
+            ScalarOperator projectExpr = postAggProjection.getValue();
+            ScalarOperator newProjectExpr = projectExpr.accept(rewriteSumVisitor, null);
+            newPostAggProjections.put(columnRefOperator, newProjectExpr);
+        }
+
+        if (rewriteSumVisitor.newAggregations.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:126]: > Could move this check early

fixed. I added FastCheckRewritableVisitor to do it

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 3 above)
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java]: > use enable instead of disable

fixed

Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java]: > for high-cardinality group-by agg, add another count agg functions, in the 1st AggregateStreamingOperator, serialized overhead would increase significantly, should we testify these scenarios, if so ,we can prevent these rewrite.

I have add a restriction, for the queries with group by keys, if we can't reduce the number of agg functions, just skip rewriting

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java (hunk 1 above)
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java]: > can we support other agg functions, for examples:
> 
> avg(v+c) = avg(v)+c avg (distinct v+c) = avg(distinct v) + c max(v+c) = max(v) +c min(v+c) = min(v) +c count(v + c) = count(v) count(distinct v+c) = count(distinct v)

yeah…I think we can rewrite them by this idea, I will do it later.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java (hunk 8 above)
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:99]: > if there is no project operator resides above agg operator, can this optimization be applied? so in such a situation, a trivial project operation always is interpolated above agg operator？

I saw that there must be a project above agg in the logical plan we obtained initially(this project may be removed in the subsequent optimization), so I think it should be no problem to use this pattern matching here.

**Comment by fzhedu** [comment]
pls add more tests that would cause overflow for integers and decimals.

can JIT also avoid such problem?

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+        OptExpression newPreAggProjectOpt = OptExpression.create(new LogicalProjectOperator(newPreAggProjections),
+                input.getInputs().get(0).getInputs().get(0).getInputs());
+
+        OptExpression newAggOpt = OptExpression.create(newAggOperator, newPreAggProjectOpt);
+
+        OptExpression newPostAggProjectOpt = OptExpression.create(newPostAggProjectOperator, newAggOpt);
+        return Lists.newArrayList(newPostAggProjectOpt);
+    }
+
+    // use this to quickly check if we can rewrite input plan
```
Comment (silverbullet233) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java:156]:
```suggestion
    // use this to quickly check if we can rewrite the aggregate function
```

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 179 / 186 (96.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java|173|180|96.11%|[117, 329, 330, 331, 332, 340, 367]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/AggregateTest.java [hunk 11]
```diff
@@ -1882,4 +1884,113 @@ public void testGroupByLiteral() throws Exception {
         assertContains(plan, "  3:Project\n" +
                 "  |  <slot 3> : -9223372036854775808");
     }
+
+    @Test
+    public void testRewriteSumByAssociativeRule() throws Exception {
+        // 1. different types
+        // 1.1 nullable
+        String sql = "select sum(t1b+1),sum(t1c+1),sum(t1d+1),sum(t1e+1),sum(t1f+1),sum(t1g+1),sum(id_decimal+1)" +
+                " from test_all_type";
```
Comment (fzhedu) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/AggregateTest.java:1893]: what about `-`,`+`,`/`?

Comment (silverbullet233) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/AggregateTest.java:1893]: > what about `-`,`*`,`/`?

* `-` is supported now. 
* I will support `*` in other PR later. 
* `/` may introduce precision problems, I need to confirm whether we can rewrite it like this

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test
